File: test/java/com/marsrovers/location/PlateauTest.java
package com.marsrovers.location;

import org.junit.Test;
import com.marsrovers.units.Rover;

import java.util.Optional;

import static junit.framework.TestCase.assertFalse;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class PlateauTest {


    @Test
    public void shouldContainBoundaries() throws Exception {

        Plateau plateau = new Plateau(10, 10);

        assertEquals(10, plateau.horizontalBoundary);
        assertEquals(10, plateau.verticalBoundary);

    }

    @Test(expected = IllegalArgumentException.class)
    public void shouldNotCreatePlateauWithoutHorizontalBoundary() throws Exception {
        new Plateau(0, 10);
    }


    @Test(expected = IllegalArgumentException.class)
    public void shouldNotCreatePlateauWithoutVerticalBoundary() throws Exception {
        new Plateau(10, 0);
    }


    @Test
    public void shouldSetANewRoverToThePlateu() throws Exception {

        Rover rover = new Rover("a", Orientation.NORTH, new Position(1,1));
        Plateau plateau = new Plateau(10, 10);

        plateau.set(rover);

        assertTrue(plateau.isOccupied(rover));
        assertEquals(rover, plateau.get("a").get());
    }

    @Test
    public void shouldAInformThatThePositionIsFree() throws Exception {
        Rover rover = new Rover("a", Orientation.NORTH, new Position(1,1));
        Plateau plateau = new Plateau(10, 10);

        plateau.set(rover);
        boolean actualResponse = plateau.isOccupied(new Rover("a", Orientation.EAST, new Position(2,2)));

        assertFalse(actualResponse);
    }

    @Test
    public void shouldAInformThatThePositionIsFreeInAnEmptyPositionsList() throws Exception {
        Plateau plateau = new Plateau(10, 10);

        boolean actualResponse = plateau.isOccupied(new Rover("a", Orientation.EAST, new Position(2,2)));

        assertFalse(actualResponse);
    }

    @Test
    public void getInexistentRoverShouldReturnEmptyOptional() throws Exception {
        Plateau plateau = new Plateau(10, 10);

        Optional<Rover> rover = plateau.get("whatever");

        assertFalse(rover.isPresent());

    }
}

File: test/java/com/marsrovers/location/PlateauTest.kt
package com.marsrovers.location

import org.junit.Test
import com.marsrovers.units.Rover

import java.util.Optional

import junit.framework.TestCase.assertFalse
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue

class PlateauTest {


    @Test
    @Throws(Exception::class)
    fun shouldContainBoundaries() {

        val plateau = Plateau(10, 10)

        assertEquals(10, plateau.horizontalBoundary.toLong())
        assertEquals(10, plateau.verticalBoundary.toLong())

    }

    @Test(expected = IllegalArgumentException::class)
    @Throws(Exception::class)
    fun shouldNotCreatePlateauWithoutHorizontalBoundary() {
        Plateau(0, 10)
    }


    @Test(expected = IllegalArgumentException::class)
    @Throws(Exception::class)
    fun shouldNotCreatePlateauWithoutVerticalBoundary() {
        Plateau(10, 0)
    }


    @Test
    @Throws(Exception::class)
    fun shouldSetANewRoverToThePlateu() {

        val rover = Rover("a", Orientation.NORTH, Position(1, 1))
        val plateau = Plateau(10, 10)

        plateau.set(rover)

        assertTrue(plateau.isOccupied(rover))
        assertEquals(rover, plateau["a"].get())
    }

    @Test
    @Throws(Exception::class)
    fun shouldAInformThatThePositionIsFree() {
        val rover = Rover("a", Orientation.NORTH, Position(1, 1))
        val plateau = Plateau(10, 10)

        plateau.set(rover)
        val actualResponse = plateau.isOccupied(Rover("a", Orientation.EAST, Position(2, 2)))

        assertFalse(actualResponse)
    }

    @Test
    @Throws(Exception::class)
    fun shouldAInformThatThePositionIsFreeInAnEmptyPositionsList() {
        val plateau = Plateau(10, 10)

        val actualResponse = plateau.isOccupied(Rover("a", Orientation.EAST, Position(2, 2)))

        assertFalse(actualResponse)
    }

    @Test
    @Throws(Exception::class)
    fun getInexistentRoverShouldReturnEmptyOptional() {
        val plateau = Plateau(10, 10)

        val rover = plateau["whatever"]

        assertFalse(rover.isPresent)

    }
}
